lib/vrl/stdlib/src/encrypt.rs

```diff
@@ -0,0 +1,340 @@
+        "AES-128-CBC-ISO10126" => encrypt_padded!(Aes128Cbc, Iso10126, plaintext, key, iv),
+        other => return Err(format!("Invalid algorithm: {}", other).into()),
+    };
+
+    Ok(Value::Bytes(Bytes::from(ciphertext)))
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Encrypt;
+
+impl Function for Encrypt {
+    fn identifier(&self) -> &'static str {
+        "encrypt"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "plaintext",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "algorithm",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "key",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "iv",
+                kind: kind::BYTES,
+                required: false,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "encrypt AES-256-CFB",
+            source: r#"encode_base64(encrypt!("data", "AES-256-CFB", key: "01234567890123456789012345678912", iv: "0123456789012345"))"#,
+            result: Ok("c/dIOA=="),
+        }]
+    }
+
+    fn compile(
+        &self,
+        _state: (&mut state::LocalEnv, &mut state::ExternalEnv),
+        _ctx: &mut FunctionCompileContext,
+        mut arguments: ArgumentList,
+    ) -> Compiled {
+        let plaintext = arguments.required("plaintext");
+        let algorithm = arguments.required("algorithm");
+        let key = arguments.required("key");
+        let iv = arguments.optional("iv");
+
+        Ok(Box::new(EncryptFn {
+            plaintext,
+            algorithm,
+            key,
+            iv,
+        }))
+    }
+
+    fn call_by_vm(&self, _ctx: &mut Context, args: &mut VmArgumentList) -> Resolved {
+        let plaintext = args.required("plaintext");
+        let algorithm = args.required("algorithm");
+        let key = args.required("key");
+        let iv = args.optional("iv");
+        encrypt(plaintext, algorithm, key, iv)
+    }
+}
+
+#[derive(Debug, Clone)]
+struct EncryptFn {
+    plaintext: Box<dyn Expression>,
+    algorithm: Box<dyn Expression>,
+    key: Box<dyn Expression>,
+    iv: Option<Box<dyn Expression>>,
+}
+
+impl Expression for EncryptFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let plaintext = self.plaintext.resolve(ctx)?;
+        let algorithm = self.algorithm.resolve(ctx)?;
+        let key = self.key.resolve(ctx)?;
+
+        let iv = self.iv.as_ref().map(|iv| iv.resolve(ctx)).transpose()?;
+
+        encrypt(plaintext, algorithm, key, iv)
+    }
+
+    fn type_def(&self, _state: (&state::LocalEnv, &state::ExternalEnv)) -> TypeDef {
+        TypeDef::bytes().fallible()
```

Comment:
Even if we have a literal algorithm name, it is still fallible since the key/iv have to be a specific size. The IV should almost never be a literal, so this will be fallible.

Reply:
I'd still support making the algorithm a literal so that Vector fails to boot if an invalid algorithm is used in the configuration.

Reply:
I'd don't see a great reason to remove the ability to specify the algorithm at runtime.
I could add a check at compile time to ensure it's a valid algorithm if a literal is given, otherwise check at runtime. Would you prefer that over only allowing a literal?

Reply:
Ah, yeah, that would be a good compromise. The only advantage I see is compile time errors if an invalid algorithm is used. I anticipate that users typically will set the algorithm statically.

Reply:
> I anticipate that users typically will set the algorithm statically
That's probably true. But I don't want to remove the possibility of using an algorithm specified in the event itself. I'll push a commit with this compromise.